Playground + new UI components

jiribuchta opened the pull request:
This pull request is adding playground functionality with new UI components which will be use for the new UI. It also changes few buttons in current UI to show how exactly it should be used.

Aiosa commented on server/php/dev_setup.php:
```diff
@@ -21,7 +21,8 @@
             require_lib("jquery");
-            require_core("env");
+            require_ui();
+            ("env");
```
What is the purpose of doing ("constant")? Is it even meaningful?

Aiosa commented on package.json:
```diff
@@ -5,6 +5,7 @@
   "main": "index.php",
   "scripts": {
     "build": "grunt all",
+    "build-css": "npx tailwindcss -i ./src/assets/tailwind-spec.css -o ./src/libs/tailwind.min.css",
```
the scrpits should also have 'dev-ui' definition that runs the UI dev setup

jiribuchta replied:
done

Aiosa commented:
Looks OK, except for a few comments. Also, I am missing README in the UI directory describing the basic usage and simple example. Or is the docs elswhere? Also check please plugins/README  modules/README  src/README and if they contain a secion on UI dev, add a short paragraph that says the UI should be developed with components, and point them to the ui/README

jiribuchta commented:
@Aiosa please check my changes:
- I have changed some readmes to have link to new UI readme
- Made grunt connect watch work, autobuild css and refresh pages
- and added new npm run dev-ui which will start connect watch

Aiosa commented on ui/components/buttons.mjs:
```diff
@@ -35,9 +36,7 @@ class Button extends BaseComponent {
-            if (options.size) this.classMap["size"] = options.size;
-            if (options.outline) this.classMap["outline"] = "btn-outline";
-            if (options.type) this.classMap["type"] = options.type;
+            this._applyOptions(options, "size", "outline", "type");
```
Please see here the change. You are programming the same logics on two places. The settings objects have built in logics, so use it. Keeping consistency == doing stuff once at one place. Try to use this function also at the collapse component.

Aiosa commented on ui/components/collapse.mjs:
```diff
@@ -3,8 +3,14 @@ import { BaseComponent } from "./baseComponent.mjs";
 
 const { div, input, details, summary } = van.tags
 
+/**
```
Missing docs!

Aiosa commented on ui/components/collapse.mjs:
```diff
@@ -3,8 +3,14 @@ import { BaseComponent } from "./baseComponent.mjs";
+ */
 class Collapse extends BaseComponent {
 
+    /**
```
Missing docs!

Aiosa commented on ui/components/fa-icon.mjs:
```diff
@@ -3,7 +3,14 @@ import { BaseComponent } from "./baseComponent.mjs";
+ * TODO
+ */
 class FAIcon extends BaseComponent {
+
```
Missing docs!

Aiosa commented on ui/test_ui.html:
```diff
@@ -139,6 +140,7 @@
             b_pickbutton.attachTo(document.getElementById("component-picker"));
 
             var b_pickCollapse = new ui.Button({
+                size: ui.Button.SIZE.SMALL,
```
This was not working properly, since you passed the option value to the classlist, but this is a FUNCTION not a STRING.

Aiosa replied:
(already fixed, just note how the `_applyOptions` method fixes & simplifies stuff)

Aiosa commented:
Please add more docs on the components. Otherwise it looks usable. The last thing I will now ignore is a dynamic component building. RN the index html contains hardcoded worskpace usage, which is unusable in the long term.

You have imported `ui` module which holds all components that have common base class. Each component defines CAPITAL-CASE properties it supports for configuration. Using these two it should be possible to build the workspace configuration dynamically. But let's leave this for a future tasks...

Aiosa commented:
@jiribuchta there still were minor issues, please see the latest PR and reflect on what was wrong there (ignore the html file changes or the new component, these are just bonusses I added when I was in the fixing process).

Now it looks like quite a well-designed UI system base. Ready, just adding a changelog..
